### PR TITLE
Fix part of #5344: Refactor Classroom List Screen UI

### PR DESCRIPTION
**Pull request by theMr17**
## Explanation

Fixes part of #5344

This PR refactors the new Classroom List Screen to align with the updated design mocks provided in [Figma](https://www.figma.com/design/H4BaHlwnALtjQcemP322hE/Oppia-Multiple-Classroom-Android-(Ashley's-Expansion)?node-id=3064-11726&t=dnQTkdXXXWoeh9k3-0). The changes include:
- Implementing functionality to collapse the classroom cards when the carousel sticks to the top.
- Updating the dark mode color scheme to match the new design specifications.

https://github.com/user-attachments/assets/cf7600a1-d0b8-4c42-af3c-67950589f8f7

https://github.com/user-attachments/assets/dfe8b564-0d37-450d-bc6a-d244e7aa58d7

### Phone
|Portrait|Landscape|
|--|--|
|image|image|
|image|image|
|image|image|
|image|image|

### Tablet
|Portrait|Landscape|
|--|--|
|image|image|
||image|
|image|image|
||image|

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only

If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

### CODE REVIEW

**Comment by theMr17** [comment]
This PR is now ready for a review, @adhiamboperes PTAL.

**Comment by adhiamboperes** [comment]
Thanks @theMr17!
Comparing the screenshots to the mocks,  I have noted the following, PTAL:
- The "select a topic to start" heading needs to be aligned vertically center.
- The classroom cards shoud have the same width as the topic cards below, so that the UI looks consistent and  makes good use of available real estate. This is especially evident in the tablet UIs, but should be updated for all layouts.
- The fontstyle used here for the "classrooms"  and "recommendation" headers is different from the mocks.

**Comment by theMr17** [comment]
> * The fontstyle used here for the "classrooms"  and "recommendation" headers is different from the mocks.

@adhiamboperes per my search, there are two ways to change fonts:
1. **Using the Fonts Directory**: Store the .ttf format fonts, which are binary files, in the project's fonts directory.
2. **Using the Downloadable Fonts API:** Fetch the required font using the Downloadable Fonts API. However, this method requires Compose version 1.2.0, while the currently supported version is 1.1.1.

Reference: https://developer.android.com/develop/ui/compose/text/fonts.

**Comment by adhiamboperes** [comment]
> > * The fontstyle used here for the "classrooms"  and "recommendation" headers is different from the mocks.
> 
> @adhiamboperes per my search, there are two ways to change fonts:
> 
> 1. **Using the Fonts Directory**: Store the .ttf format fonts, which are binary files, in the project's fonts directory.
> 2. **Using the Downloadable Fonts API:** Fetch the required font using the Downloadable Fonts API. However, this method requires Compose version 1.2.0, while the currently supported version is 1.1.1.
> 
> Reference: https://developer.android.com/develop/ui/compose/text/fonts.

@theMr17, font style attributes are part of the Text object. To change the font style you only need to apply any of the following to the textview, that are relevant.
```
 fontFamily = FontFamily.SansSerif,
fontWeight = FontWeight.Medium,
fontStyle = FontStyle.Italic,
 ``` 
 
 Edit: The mocks have a lighter font. The color is silver instead of white as in the screenshots, and the size/weight might also be different. Please verify.

**Comment by theMr17** [comment]
@adhiamboperes PTAL.

File: app/src/main/java/org/oppia/android/app/classroom/classroomlist/ClassroomList.kt [hunk 1]
```diff
@@ -140,3 +129,17 @@ fun ClassroomCard(
     }
   }
 }
+
+@Composable
```
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/classroom/classroomlist/ClassroomList.kt:133]: This is not a composable, so there is no need to annotate it as such.

File: app/src/main/java/org/oppia/android/app/home/classroomlist/AllClassroomsViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package org.oppia.android.app.home.classroomlist
+
+import androidx.lifecycle.ViewModel
+import org.oppia.android.app.classroom.ClassroomListFragment
+import org.oppia.android.app.home.HomeItemViewModel
+
+/** [ViewModel] all classrooms text in [ClassroomListFragment]. */
```
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/home/classroomlist/AllClassroomsViewModel.kt]:
```suggestion
/** [ViewModel] for displaying the classroom header in [ClassroomListFragment]. */
```

File: app/src/main/java/org/oppia/android/app/classroom/ClassroomListFragmentPresenter.kt [hunk 3]
```diff
@@ -197,15 +200,20 @@ class ClassroomListFragmentPresenter @Inject constructor(
       ?.plus(classroomListViewModel.topicList)
       ?.groupBy { it::class }
     val topicListSpanCount = integerResource(id = R.integer.home_span_count)
+    val listState = rememberLazyListState()
+    var classroomListIndex = 0
```
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/classroom/ClassroomListFragmentPresenter.kt]: I wasn't able to understand why we are tracking an index in this file. Could you please clarify?

**Review by adhiamboperes — CHANGES_REQUESTED** [review]
Thanks @theMr17. PTAL at the comments inline.

**Comment by adhiamboperes** [comment]
@seanlip, could you PTAL at the screenrecording/updated screenshots wrt to M1 product demo feedback on design?

File: app/src/main/java/org/oppia/android/app/classroom/classroomlist/ClassroomList.kt (hunk 1 above)
Comment (theMr17) [reply app/src/main/java/org/oppia/android/app/classroom/classroomlist/ClassroomList.kt:133]: The following functions and variables can only be called or accessed within a `@Composable` scope:
- `LocalConfiguration.current`
- `dimensionResource`
- `integerResource`

Therefore, this function is annotated with `@Composable`.

File: app/src/main/java/org/oppia/android/app/home/classroomlist/AllClassroomsViewModel.kt (hunk 2 above)
Comment (theMr17) [reply app/src/main/java/org/oppia/android/app/home/classroomlist/AllClassroomsViewModel.kt]: Done.

File: app/src/main/java/org/oppia/android/app/classroom/ClassroomListFragmentPresenter.kt (hunk 3 above)
Comment (theMr17) [reply app/src/main/java/org/oppia/android/app/classroom/ClassroomListFragmentPresenter.kt]: The classroom carousel can be at either index 1 or 2, depending on the visibility of the promoted stories (which in turn depends on the profile login count).

Once the index of the classroom carousel is determined, `listState.firstVisibleItemIndex >= classroomListIndex` is checked to determine if the classroom carousel is sticky at the top. If the carousel is sticky, the `isSticky` parameter is set to `true` ([here](https://github.com/oppia/oppia-android/blob/043628ba8884d93b6ae394939f354d978f6c6d4d/app/src/main/java/org/oppia/android/app/classroom/ClassroomListFragmentPresenter.kt#L245)) and is used to collapse the classroom cards.

**Comment by theMr17** [comment]
Addressed the review comments, @adhiamboperes PTAL.

**Comment by seanlip** [comment]
> @seanlip, could you PTAL at the screenrecording/updated screenshots wrt to M1 product demo feedback on design?

@adhiamboperes Thanks for looping me in. I took a look and have a few questions (/cc @theMr17 )

- I can't actually tell which of the topics at the bottom are supposed to be in the Science vs the Math classroom, since they all seem to relate to Math. So I can't tell which topics are meant to correspond to which tab. My sense is that First Test Topic etc. are in the Maths classroom because the background of that tab is the same as the background of the part behind the topic cards. But I'm not totally sure about this because the outlining on the Science/Math header seems to contradict it (it sort of implies that Math is clickable). So I'm a bit confused overall.
- The animation of the Science/Math headers to lose their images when their height is shortened looks really nice!
- Would it be possible to see what this looks like in light mode as well? (I was referring to the [M1 comments](https://docs.google.com/document/d/14Z6EJXYgic9QR0MCnnoWqeVq5etpfVxwj-RznC_vj1Y/edit) and I don't see a note about them being dark-mode only.)

Also @ashley-ya Could you please take a look at the videos/screenshots above and see if you have any thoughts on the UI (cf my notes here too)? Thanks!

**Comment by theMr17** [comment]
> * I can't actually tell which of the topics at the bottom are supposed to be in the Science vs the Math classroom, since they all seem to relate to Math. So I can't tell which topics are meant to correspond to which tab. My sense is that First Test Topic etc. are in the Maths classroom because the background of that tab is the same as the background of the part behind the topic cards. But I'm not totally sure about this because the outlining on the Science/Math header seems to contradict it (it sort of implies that Math is clickable). So I'm a bit confused overall.

Hi @ashley-ya,  Seems like this needs a design decision, PTAL.

> * Would it be possible to see what this looks like in light mode as well? (I was referring to the [M1 comments](https://docs.google.com/document/d/14Z6EJXYgic9QR0MCnnoWqeVq5etpfVxwj-RznC_vj1Y/edit) and I don't see a note about them being dark-mode only.)

@seanlip Thanks for the feedback! Updated the PR description with light mode screen recording and screenshots, PTAL.

**Comment by ashley-ya** [comment]
> > @seanlip, could you PTAL at the screenrecording/updated screenshots wrt to M1 product demo feedback on design?
> 
> @adhiamboperes Thanks for looping me in. I took a look and have a few questions (/cc @theMr17 )
> 
>     * I can't actually tell which of the topics at the bottom are supposed to be in the Science vs the Math classroom, since they all seem to relate to Math. So I can't tell which topics are meant to correspond to which tab. My sense is that First Test Topic etc. are in the Maths classroom because the background of that tab is the same as the background of the part behind the topic cards. But I'm not totally sure about this because the outlining on the Science/Math header seems to contradict it (it sort of implies that Math is clickable). So I'm a bit confused overall.
> 
>     * The animation of the Science/Math headers to lose their images when their height is shortened looks really nice!
> 
>     * Would it be possible to see what this looks like in light mode as well? (I was referring to the [M1 comments](https://docs.google.com/document/d/14Z6EJXYgic9QR0MCnnoWqeVq5etpfVxwj-RznC_vj1Y/edit) and I don't see a note about them being dark-mode only.)
> 
> 
> Also @ashley-ya Could you please take a look at the videos/screenshots above and see if you have any thoughts on the UI (cf my notes here too)? Thanks!


Hi, I'll take a look!

File: app/src/main/java/org/oppia/android/app/classroom/classroomlist/AllClassroomsHeaderText.kt [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import androidx.compose.ui.platform.testTag
+import androidx.compose.ui.res.colorResource
+import androidx.compose.ui.res.dimensionResource
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.text.font.FontFamily
+import androidx.compose.ui.text.font.FontWeight
+import androidx.compose.ui.unit.sp
+import org.oppia.android.R
+
+/** Test tag for the all topics section header. */
```
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/classroom/classroomlist/AllClassroomsHeaderText.kt]:
```suggestion
/** Test tag for the all classrooms section header. */
```

File: app/src/main/java/org/oppia/android/app/classroom/classroomlist/AllClassroomsHeaderText.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+const val ALL_CLASSROOMS_HEADER_TEST_TAG = "TEST_TAG.all_classrooms_header"
+
+/** Displays the header text for the topic list section. */
```
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/classroom/classroomlist/AllClassroomsHeaderText.kt]:
```suggestion
/** Displays the header text for the classrooms list section. */
```

File: app/src/main/java/org/oppia/android/app/classroom/ClassroomListFragmentPresenter.kt (hunk 3 above)
Comment (adhiamboperes) [reply app/src/main/java/org/oppia/android/app/classroom/ClassroomListFragmentPresenter.kt]: I feel that manually managing the index is brittle and error prone. `LazyColumn` has `itemsIndexed` to manage the index of items dynamically without the need for manual indexing. You can switch to this instead.
```kotlin
isSticky = listState.firstVisibleItemIndex >= items.indexOfFirst { it is ClassroomSummaryViewModel }
```

**Review by adhiamboperes — CHANGES_REQUESTED** [review]
Thanks @theMr17! I have left followup comments.

**Comment by adhiamboperes** [comment]
> * I can't actually tell which of the topics at the bottom are supposed to be in the Science vs the Math classroom, since they all seem to relate to Math. So I can't tell which topics are meant to correspond to which tab. My sense is that First Test Topic etc. are in the Maths classroom because the background of that tab is the same as the background of the part behind the topic cards. But I'm not totally sure about this because the outlining on the Science/Math header seems to contradict it (it sort of implies that Math is clickable). So I'm a bit confused overall.


@seanlip, this feedback isn't very clear to to me. Are you making a reference to the darkmode designs or both? And by outlining the header, do you mean the card/tab border? 

I believe the goal is to have the background of the selected classroom card to be the same as the topic list background, and this is apparent in the lightmode screenshots. However, in the darkmode ones, the colors seem to be switched such that the unselected card is the one that matches. @theMr17, could you please review this?

File: app/src/main/java/org/oppia/android/app/classroom/classroomlist/AllClassroomsHeaderText.kt (hunk 4 above)
Comment (theMr17) [reply app/src/main/java/org/oppia/android/app/classroom/classroomlist/AllClassroomsHeaderText.kt]: Done.

File: app/src/main/java/org/oppia/android/app/classroom/classroomlist/AllClassroomsHeaderText.kt (hunk 5 above)
Comment (theMr17) [reply app/src/main/java/org/oppia/android/app/classroom/classroomlist/AllClassroomsHeaderText.kt]: Done.

File: app/src/main/java/org/oppia/android/app/classroom/ClassroomListFragmentPresenter.kt (hunk 3 above)
Comment (theMr17) [reply app/src/main/java/org/oppia/android/app/classroom/ClassroomListFragmentPresenter.kt]: Updated with an approach similar to the suggestion.

**Comment by theMr17** [comment]
> Thanks @theMr17! I have left followup comments.

@adhiamboperes PTAL, addressed the comments.

> I believe the goal is to have the background of the selected classroom card to be the same as the topic list background, and this is apparent in the lightmode screenshots. However, in the darkmode ones, the colors seem to be switched such that the unselected card is the one that matches. @theMr17, could you please review this?

You are right, I have updated the selected classroom card colour to match the topic list background in dark mode as well.

**Comment by seanlip** [comment]
> > * I can't actually tell which of the topics at the bottom are supposed to be in the Science vs the Math classroom, since they all seem to relate to Math. So I can't tell which topics are meant to correspond to which tab. My sense is that First Test Topic etc. are in the Maths classroom because the background of that tab is the same as the background of the part behind the topic cards. But I'm not totally sure about this because the outlining on the Science/Math header seems to contradict it (it sort of implies that Math is clickable). So I'm a bit confused overall.
> 
> @seanlip, this feedback isn't very clear to to me. Are you making a reference to the darkmode designs or both? And bu outlining the header, do you mean the card/tab border?

Hi @adhiamboperes, sorry for being unclear. Yes, I think your interpretation of my feedback is correct overall. I was initially referring to the dark mode designs since they were the only ones available at the time I looked at this PR. I will take a look at the updated videos now.

**Comment by seanlip** [comment]
@theMr17 @adhiamboperes OK, here are my remaining comments/concerns (@ashley-ya it would be helpful to get your feedback on these as well!):

- For light mode, I am wondering if we need the bottom border that separates the "Science" header tab from the set of topics. It makes it feel like the Science header is not related to the "Select a Topic to Start" list at the bottom. I wonder if removing just the bottom borders of both the "Science" and "Maths" tabs would look OK and help with that, so that the colour of the tab is continuous with the colour of the "list of topics" background.
- Somehow this is less prominent in dark mode (and I think the colouring works correctly now).

Otherwise, no issues from my end. Thanks!

**Review by adhiamboperes — APPROVED** [review]
Thanks @theMr17! Code changes LGTM.

**Comment by theMr17** [comment]
> * For light mode, I am wondering if we need the bottom border that separates the "Science" header tab from the set of topics. It makes it feel like the Science header is not related to the "Select a Topic to Start" list at the bottom. I wonder if removing just the bottom borders of both the "Science" and "Maths" tabs would look OK and help with that, so that the colour of the tab is continuous with the colour of the "list of topics" background.

Hi @seanlip, per suggestion, I tried to remove the bottom border, but I realized that Jetpack Compose does not support that. We cannot add borders only to the top, start, and end; we can either add borders to all sides or not add them at all.

**Comment by ashley-ya** [comment]
Hi @theMr17 great work on the multiple classrooms tab and animation! A few tweaks to the UI to improve the interface clarity:

1. the new light mode selected tab background color will be HEX# CFE7E5. it should be darker than the topics background, and will help users understand that its the selected classroom.
2. I've realized the maths icon gets lost in the dark background. I've created a new SVG for the dark mode, attached here.
math Icon - Dark Mode

It can also be found in the figma file: https://www.figma.com/design/H4BaHlwnALtjQcemP322hE/Oppia-Multiple-Classroom-Android-(Ashley's-Expansion)?node-id=3168-3352&t=te0J1N9h83aZb9PK-1

Thanks again for your work on this!

**Comment by ashley-ya** [comment]
> @theMr17 @adhiamboperes OK, here are my remaining comments/concerns (@ashley-ya it would be helpful to get your feedback on these as well!):
> 
>     * For light mode, I am wondering if we need the bottom border that separates the "Science" header tab from the set of topics. It makes it feel like the Science header is not related to the "Select a Topic to Start" list at the bottom. I wonder if removing just the bottom borders of both the "Science" and "Maths" tabs would look OK and help with that, so that the colour of the tab is continuous with the colour of the "list of topics" background.
> 
>     * Somehow this is less prominent in dark mode (and I think the colouring works correctly now).
> 
> 
> Otherwise, no issues from my end. Thanks!

Hi @theMr17 @seanlip ,

Apologies for the delay. I see your point of moving towards a design that clearly connects the classrooms and their related topics. The issue I see is that once the user selects a classroom, the selected tab background and topics section background wouldn't be the same color and the lack of border may appear strange. 

With the information Saptak shared about the border options, I think the design might need more UI rework to see how it could function without any tab borders.

So I'm proposing we keep the style for now

**Comment by theMr17** [comment]
Thanks @ashley-ya for the new dark mode icon. Moving ahead with the same style for now!

**Comment by theMr17** [comment]
@adhiamboperes PTAL at the latest code changes and the updated screenshots and recordings.

**Comment by seanlip** [comment]
Thanks @ashley-ya! 

@theMr17 @adhiamboperes FYI I'm happy with this getting merged as long as @ashley-ya's comments are addressed.

**Review by adhiamboperes — APPROVED** [review]
Thanks @theMr17! This LGTM.

**Pull request merged**